plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,195 @@
+            type: str
+
+    requirements:
+    - "etcd3 >= 0.10"
+'''
+
+EXAMPLES = '''
+    - name: "a value from a locally running etcd"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar') }}"
+
+    - name: "values from multiple folders on a locally running etcd"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo', 'bar', 'baz') }}"
+
+    - name: "look for a key prefix"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', '/foo/bar', prefix=True) }}"
+
+    - name: "connect to etcd3 with a client certificate"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar', cert_cert='/etc/ssl/etcd/client.pem', cert_key='/etc/ssl/etcd/client.key') }}"
+'''
+
+RETURN = '''
+    _raw:
+        description:
+        - List of keys and associated values.
+        type: list
+        elements: complex
+'''
+
+import re
+import yaml
+
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+from ansible.plugins.lookup import LookupBase
+from ansible.errors import AnsibleLookupError
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+display = Display()
+
+
+def etcd3_client(client_params):
+    try:
+        etcd = etcd3.client(**client_params)
+        etcd.status()
+    except Exception as exp:
+        raise AnsibleLookupError('Cannot connect to etcd cluster: %s' % (to_native(exp)))
+    return etcd
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+        ymldoc = yaml.safe_load(DOCUMENTATION)
+
+        if not HAS_ETCD:
+            display.error(missing_required_lib('etcd3'))
+            return None
+
+        # create the etcd3 connection parameters objet to pass to etcd3 class
+        # by copying options object and removing unneeded keys
+        client_params = dict(self._options)
+        for i in ('_terms', 'prefix', 'endpoints'):
+            if i in client_params:
+                client_params.pop(i)
+
+        # etcd3 class expects host and port as connection parameters, so endpoints
+        # must be mangled a bit to fit in this scheme.
+        # so here we use a regex to extract server and port
+        if self.get_option('endpoints'):
+            match = re.compile(
+                r'^(https?://)?(?P<host>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|([-_\d\w\.]+))(:(?P<port>\d{1,5}))?/?$'
+            ).match(self.get_option('endpoints'))
+            if match:
+                if match.group('host') and match.group('host') != ymldoc['options']['host']['default']:
+                    client_params['host'] = match.group('host')
+                    display.verbose("etcd3 option 'host' overriden by 'endpoint' host value: %s" % client_params['host'])
```

Comment:
This precedence must be documented. Otherwise it is completely surprising to users. I would expect that if I explicitly set `host` to a value, and for some reason the environment variable `ETCDCTL_ENDPOINTS` happened to be set, that my `host` value will be used and **not** the value from `ETCDCTL_ENDPOINTS`.

Maybe it is better to drop `host` and `port` and only allow `endpoints` to specify the host(s) to use. That would simplify the logic and prevent unexpected behavior.

Reply:
I see your point of view, but I wanted to stay as soon as possible of the etcd3 module parameters. The point is, IMHO, etcd3 users expects to use "endpoints" as it is the "natural" approach for experienced etcd users. But the etcd3 python module doesn't use endpoints but ip addr and port.

I think it is important to propose both but you're right: host/port and endpoints should be mutually exclusive and raise an error if both used at same time.

Reply:
How about removing the defaults from `host` and `port`, instead add a default to `endpoints`, and state that `host` and `port` can be used to overwrite the values extracted from `endpoints`? I think that would be a lot cleaner, and you also wouldn't have to fiddle with Ansible's option parsing as you're currently still doing (https://github.com/ansible-collections/community.general/pull/127/commits/17788d0c3c13256840f452ce1380ba2cb98e7399#diff-80143e5a3a0d857a1db54bc5c02162a9R181).

Reply:
I agree with your proposal of handling default value to `endpoint` instead of `host` and `port`, this is actually a better idea ;)
But I still need to handle mutual exclusion. Do you have a cleaner approach to suggest than checking `kwargs` ?
This is maybe not the most _pythonic_ way to address this but browsing the existing core I did not found any method that whould tell me if an option is actually passed as a parameter or not.

Reply:
I don't think you need to handle mutual exclusion. I would simply state in the docs that `host` and `port` take precedence over values parsed from `endpoints`.

I'm not sure whether simply checking `kwargs` or calling `C.config.get_config_value_and_origin()` is cleaner. I'd think that using none, and simply defining a simle precedence schema and documenting it should be enough.

If you really want to keep the current behavior, please document it properly - `endpoints` itself (with the possibility to use env variables and default) is NOT mutually exclusive to `host`+`port`, but only specifying `endpoints` directly is mutually exclusive to `host`+`port`.

Reply:
Hem... I pushed a commit _before_ reading your comment. Sorry ! So if I remove lines 181-182
```python
        if 'endpoints' in kwargs.keys() and 'host' in kwargs.keys() or 'port' in kwargs.keys():
            raise AnsibleError("etcd3 lookup option 'endpoints' is mutually exclusive with 'host' and 'port' options !")
```
It will work as you describes: `host` and `port` are first extracted from endpoint, then eventually overriden by `host` and `port` options if they are defined by user.

Please also note I added metadata and *notes* and *seealso* sections in docstring. I hope I did not introduced something wrong...

Reply:
Do whatever you prefer :) BTW, `'endpoints' in kwargs` is more pythonic and efficient than `'endpoints' in kwargs.keys()`.

Reply:
Actually I finally prefer your idea of handling options precedence. I feel it more elegant. And thanks for the tips on dict ;)